### DISCUSSION TITLE
Accessing FormBuilderState currentState value within `dispose`

### DISCUSSION THREAD

**Discussion by judsonjames** [danvick/flutter_form_builder#1260] [Q&A] [answered]
I'm using Flutter Form Builder in order to collect data, then save to my database using Isar. When a user pops the widget either with the OS dependent "back button" or the "back button" on the AppBar, I want to store the state of the Form to a draft value. However, also I'm utilizing GoRouter, which [doesn't work with `WillPopScope`](https://github.com/flutter/flutter/issues/99113). It works fine on my desktop emulator, but on a native Android device it will fail to pop the widget with either back button. So, given a Widget that looks like this:

```dart
class _MyWidgetState extends State<MyWidget> {
  final _formKey = GlobalKey<FormBuilderState>();
  late Isar _isar;

  @override
  void initState() {
    super.initState();
    _isar = context.read<IsarModel>().isar;
  }
  
  // I know this doesn't work, but this is ideally what I'd like to use
  @override
  void dispose() {
    if (_formKey.currentState?.value != null) {
      // write values from Form to Isar entry
    }
  }
  
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: FormBuilder(
        key: _formKey,
        body: Column(
          // FormBuilderFields
        )
      );
  }
}
```

I can't get the form value because the widget has already popped from the tree and doesn't exist. I can't use the `willPopScope` property of `FormBuilder` because it conflicts with `GoRouter`. I've tried using `routerDelegate` to get the data, but it gives inconsistent results from FormState depending on what fields were populated.

I've considered using `Provider` to cache the data of the Form from the `onChanged` callback on `FormBuilder` but I think that's overkill and could be a performance hit. Are there any other alternatives native to FlutterFormBuilder that allow the FormState to be accessed before the widget is popped?

**Comment by deandreamatias** [comment, marked as answer]
Hi, sorry for delay response
Maybe you can use `onChange` property of FormBuilder to save values when changes.
Remember to use `_formKey.currentState?.instantValue` instead `_formKey.currentState?.value`, like I explain [here](https://github.com/flutter-form-builder-ecosystem/flutter_form_builder#validate-and-get-values).
Also, the getters `_formKey.currentState?.isDirty` and `_formKey.currentState?.isTouched` can be useful
